### DISCUSSION TITLE
Best strategy to handle polymorphic relations

### DISCUSSION THREAD

**Discussion by famschopman** [ravendb/ravendb#13418] [Q&A] [answered]
I have multiple implementations of a connection that are saved with their own CLR. With a multi map index I can query them all at once, great.

MSSQLConnection inherits Connection (IConnection)
RedshiftConnection inherits Connection (IConnection)
SnowflakeConnection inherits Connection (IConnection)

On saving, we also save a schema as a separate document (describing selected tables and columns). In this schema document I have a property called ConnectionId that is referencing the ID of either of these implementations.

So it can be "ConnectionId": "MSSQLConnection-417-A" or "ConnectionId": "RedshiftConnection-417-A", or "ConnectionId": "SnowflakeConnection-417-A",

Now I want to query a schema, and then load the relevant connection with it. Unfortunately, there is no way to query documents only by ConnectionId, and then lookup the CLR type in the RavenDB document to cast it back to the implementation (right?).

The other option is to create different properties for each of the implementations in the schema document. 

What would be the best approach to handle this? It looks like a fairly common use case?

**Comment by famschopman** [comment]
Played a bit around, it is not a beauty and senior developers will call me a fool for doing this, but it works. The assembly name could be looked up with a bit of reflection from the appdomain.

        public async Task<IHttpActionResult> Post([FromBody] MSSQLConnectionViewModel model)
        {            
            // persist connection
            var connection = model.ToModel();
            await _documentSession.StoreAsync(connection);

            // persist schema, and reference connection (could be connection, table, worksheet or view implementation of a datasource)
            var schema = model.Schema;
            schema.DatasourceId = connection.Id;
            await _documentSession.StoreAsync(schema);

            // quick n dirty test: retrieve connection and cast back to concrete class
            dynamic document = _documentSession.LoadAsync<dynamic>(schema.DatasourceId);
            var typeName = document.Result.ToString(); // gives me back "Namespace.Classname"
            Type type = Type.GetType(String.Format("{0}, {1}", typeName, "AssemblyName")); // get the concrete type                    
            var jsonData = JsonConvert.SerializeObject(document);
            var connection = JsonConvert.DeserializeObject(jsonData, type);

**Comment by haludi** [comment]
Hi,

You can just load it to an object.

var document = await _documentSession.LoadAsync<objet>(schema.DatasourceId);

Be aware, you are counting on the fact RavenDB client stores the C# type in
the document metadata.
[image: Screenshot from 2022-01-10 09-53-14.png]
That is why your code works (and my suggestion).
If you have a document from non-C# client you will have a problem.
The recommended way is to decide the deserialized type based on the
document content.
The easiest way is to have an enum (or string) property dedicated for that
purpose.

Best regards,
Igal

**Reply by famschopman** [reply]
The object is not known runtime, so the only reference I have is basically the document itself (that is why I load the document as a dynamic initially) and then try to cast / deserialize it back to the corresponding type.

It just feels like I am doing the wrong things for a pretty common use case.

**Comment by haludi** [comment]
You got the type because RavenDB client is using the metadata.
You are relying on a metadata property that the RavenDB C# client adds.
That way you got the type name from document.Result.ToString().
It was already deserialized to the right object.
Then you do serialize and deserialize with no need.

**Reply by famschopman** [reply]
Yup, I am fully relying on the CLR metadata here. I could have stored a Connection with a ConnectionType property, and then deserialize during runtime based on the ConnectionType to the appropriate implementation of Connection (e.g. MSSQLConnection::Connection, or SnowflakeConnection::Connection).

What would loading the object versus dynamic improve for my use case? In both cases I would have to cast/deserialize back to the actual implementation. I also tried something cheeky along the line of .LoadAsync<IConnection>(schema.DatasourceId); which would have enabled easier casting to the concrete after but it really required a concrete type.

**Comment by haludi** [comment, marked as answer]
You can use dynamic but in your code, you loaded the `Task` to dynamic (not
the result).
You should add the `await` keyword.
Anyway, the result of the task already contains the deserialized object.
Just don't serialize and deserialize the object.

**Reply by famschopman** [reply]
I think I am getting you now. 

I probably missed the await and the result was not yet calculated. This basic scenario works. Casting it to an IAnimal enables me to work with it.

```
public interface IAnimal{}
public abstract class Animal : IAnimal{
    public string Id { get; set; }
}
public class Dog : Animal{
    public string Sound { get; set; } = "barks";
}
public class Cat : Animal
{
    public string Sound { get; set; } = "miaow";
}
public class Owner {
    public string Id { get; set; }
    public string AnimalId { get; set; }
}
```

        
```
// variant a
var dog = new Dog();
await _documentSession.StoreAsync(dog);

var dog_owner = new Owner();
dog_owner.AnimalId = dog.Id;
await _documentSession.StoreAsync(dog_owner);


// variant b            
var cat = new Cat();
await _documentSession.StoreAsync(cat);

var cat_owner = new Owner();
cat_owner.AnimalId = cat.Id;
await _documentSession.StoreAsync(cat_owner);

// now, based on the owner, get the right animal back based just using the Animal.Id
IAnimal document1 = await _documentSession.LoadAsync<dynamic>(dog_owner.AnimalId); // returns Dog
var document2 = await _documentSession.LoadAsync<object>(cat_owner.AnimalId) as IAnimal; // returns Cat
```

**Reply by ayende** [reply]
You can also use:

`var document1 = await _documentSession.LoadAsync<IAnimal>(dog_owner.AnimalId); // returns Dog`

Or:

`var document2 = await _documentSession.LoadAsync<Animal>(cat_owner.AnimalId); // returns Cat`

**Reply by ayende** [reply]
No need for `dynamic`

**Reply by mcrio** [reply]
> The recommended way is to decide the deserialized type based on the document content. The easiest way is to have an enum (or string) property dedicated for that purpose.

This seems safer than relying on the c# client metadata and the fact the document class will not move to another namespace. In this scenario, would we load the document as `object` and do a serialize, unserialize like in the example above?

**Reply by ayende** [reply]
There is the `FindClrType` in the Conventions if you want to use this sort of behavior, but honestly, until you need to, just use the metadata.

**Reply by famschopman** [reply]
Great feedback so far.

Aren't we overstating the impact of relying on CLR information? I actually considered adding a type property on the base class but actually asked myself why, if we have the metadata available. I mean, just switching from RavenDB to a different solution is one of those things often mentioned but in reality uncommon as you build your entire business logic around the features and capabilities of the platform.

If you do want to change the namespace you can perform a patch on the documents.

**Reply by ayende** [reply]
If you modify the namespace, you can also do re-routing to the right type using `FindClrTypeName` in the conventions

**Reply by mcrio** [reply]
> If you do want to change the namespace you can perform a patch on the documents.

I was more worried about a scenario when there is an accidental namespace modification during refactoring that would not be caught in integration tests that do not operate on preexisting data.

Nice to have `FindClrTypeName `.
